docs/design/proposals/rbd-mirror.md

```diff
@@ -0,0 +1,73 @@
+corresponding PV name and the unique ID mapping, It creates the RBD image with
+the unique ID and returns the encoded value which contains all the required
+information for other operations. For mirroring, the same RBD image will be
+mirrored to the secondary cluster. As the journal(omap data) is not mirrored to
+the secondary cluster, The cephcsi cannot identify the RBD image corresponding
+to the PV wihtout omap data.
+
+**Note:** It's expected that the user/admin will create the static PVC's on the
+secondary site during the failover.
+
+**Note:** when the static PVC created on the secondary site we cannot use the
+volumeHandle to identify the omap data or the image anymore because as we have
+only PoolID and clusterID in the volumeHandle. we cannot identify the correct
+pool name from the poolID because pool name will remain the same on both clusters
+but not the poolID even the clusterID can be different on the secondary
+cluster.
+
+> **VolumeHandle** is the unique volume name returned by the CSI volume plugin’s
+CreateVolume to refer to the volume on all subsequent calls.
+
+Once the static PVC is created on the secondary cluster, the user can try
+delete the PVC,expand the PVC or mount the PVC. Incase of mounting (node stage)
+we will get the volume context in GRPC call but not in the Delete/Expand
+Request. In Delete/Expand GRPC request only the
+volumeHandle (`clusterID-poolID-volumeuniqueID`) will be sent where it contains
+the encoded information related to clusterID and poolID. The volumeHandle is
+not useful in secondary cluster as the poolID and clusterID always may not be
+the same.
+
+To solve this problem, We will have a new controller running as part of
+provisioner pod which watchers for the PV objects. When a PV is created it will
+extract the required information from the PV spec and it will regenerate the
+omap data and also it will generate a new volumeHandle
+(`newclusterID-newpoolID-volumeuniqueID`) and it creates an omap object for
+mapping between old volumeHandle and new volumeHandle. whenever cephcsi gets a
+GRPC request with older volumeHandle it will check is there any new
+volumeHandle exists for the old volumeHandle. If yes, it uses the new
+volumeHandle for internal operations (to get pool name, monitor details from
+the clusterID etc).
+
+Currently, we are making use of watchers in node stage request to make sure RWO
+PVC is mounted on a single node at a given point in time. We need to change the
+watchers logic in the node stage request as when we enable the RBD mirroring on
+an image, a watcher will be added on a RBD image by the rbd mirroring daemon.
+
+To solve the clusterID problem, If the clusterID is different on the second
+cluster the admin as to create a new configmap for the mapped clusterID's.
+whenever cephcsi gets a request it will check the clusterID mapping exists and
+uses mapped clusterID to get the information like monitors etc.
+
+**This design does not cover the below items:**
```

Comment:
Are there any other requirements to the environment, like firewalls, response times between Ceph clusters, ...
As steps to do failover/fallback are not covered, maybe explain that *automatic failover* will not be available (yet) either.
What other pieces are missing? Integration with other tools that sync objects between Kubernetes clusters?
This only discusses storage, pointers/links to how applications can be mirrored would be nice to have. No details, just enough to get an idea how a complete setup could be achieved at one point.

Reply:
> Are there any other requirements to the environment, like firewalls, response times between Ceph clusters, ...

Not planning to talk about mirror bootstrapping etc, maybe this can come as part of the usage document

> As steps to do failover/fallback are not covered, maybe explain that _automatic failover_ will not be available (yet) either.

step to do failover will be covered in usage document

> What other pieces are missing? Integration with other tools that sync objects between Kubernetes clusters?

Its kube admin responsibility to take application backup from one cluster and restore it on another cluster (from cephcsi point of view the PVC and PV objects should remain same)

> This only discusses storage, pointers/links to how applications can be mirrored would be nice to have. No details, just enough to get an idea how a complete setup could be achieved at one point.

We can add this to usage doc.